chips/e310x/src/chip.rs

```diff
@@ -136,16 +136,142 @@ impl kernel::Chip for E310x {
     }
 }
 
+pub unsafe fn handle_trap() {
+    let cause = rv32i::riscvregs::register::mcause::read();
+    // if most sig bit is set, is interrupt
+    if cause.is_interrupt() {
+        // strip off the msb
+        match rv32i::riscvregs::register::mcause::Interrupt::from(cause.code()) {
+            rv32i::riscvregs::register::mcause::Interrupt::MachineSoft => {
+                debug!("encountered machine mode software interrupt");
+            }
+            // should never occur
+            rv32i::riscvregs::register::mcause::Interrupt::UserSoft => (),
+            rv32i::riscvregs::register::mcause::Interrupt::SupervisorSoft => (),
+
+            rv32i::riscvregs::register::mcause::Interrupt::MachineTimer => (),
+            // should never occur
+            rv32i::riscvregs::register::mcause::Interrupt::UserTimer => (),
+            rv32i::riscvregs::register::mcause::Interrupt::SupervisorTimer => (),
+
+            // this includes UART, GPIO pins, etc
+            rv32i::riscvregs::register::mcause::Interrupt::MachineExternal => {
+                // just send out a message that the interrupt occurred and complete it
+                let ext_interrupt_wrapper = plic::next_pending();
+                match ext_interrupt_wrapper {
+                    None => (),
+                    Some(ext_interrupt_id) => {
+                        debug!("interrupt triggered {}\n", ext_interrupt_id);
+                        plic::complete(ext_interrupt_id);
```

Comment:
This doesn't look quite right. The trap handler should only disable the interrupt, not mark it as finished. Or does this not work how I think?

Reply:
A couple of things to note here. Spec is best described in section 10.7 [here](https://sifive.cdn.prismic.io/sifive%2F4d063bf8-3ae6-4db6-9843-ee9076ebadf7_fe310-g000.pdf). It's under-specified by the spec what happens if you claim the interrupt (read the claim/complete MMIO) multiple times. So at the very least I don't think we can rely on being able to handle the interrupt anytime after this. On the e310 board, from the testing I've done (for example, add in another claim before completing in this same spot), you get interrupt id 0 (no more pending interrupts). So we either need to claim, handle the interrupt, and complete the interrupt here, or have some global "in progress" state that we write to and then read from in service_pending_interrupts to complete. This seemed easier at the time. With this added knowledge, do you think it's worth changing? I can probably get a change out really quickly.

You're right, I should disable the interrupt here. 

Ideally, all interrupts should be suppressed after the reset_handler, so my intention was just to notify that the interrupt occurred (as it's not expected behavior), and then ignore it.